jetty-util/src/main/java/org/eclipse/jetty/util/ssl/X509.java

```diff
@@ -182,21 +182,45 @@ public boolean matches(String host)
                 return true;
         }
 
-        InetAddress address = toInetAddress(host);
-        if (address != null)
-            return _addresses.contains(address);
+        // Check if the host looks like an IP address to avoid
+        // DNS lookup for host names that did not match.
+        if (seemsIPAddress(host))
+        {
+            InetAddress address = toInetAddress(host);
+            if (address != null)
+                return _addresses.contains(address);
+        }
 
         return false;
     }
 
+    private static boolean seemsIPAddress(String host)
+    {
+        // IPv4 is just numbers and dots.
+        String ipv4RegExp = "[0-9\\.]+";
+        // IPv6 is hex and colons and possibly brackets.
+        String ipv6RegExp = "[0-9a-fA-F:\\[\\]]+";
+        return host.matches(ipv4RegExp) ||
+            (host.matches(ipv6RegExp) && containsAtLeastTwoColons(host));
```

Comment:
I missed this before the port... but shouldn't we precompile these regex's? 
Also they can be more precise: IPv4 must start and finish with a digit; IPv6 may have [] only on the outside and must have a : somewhere (which can be done in regex rather than another method).   i think this is specially important for ipv6 matching as `a-fA-F` can spell out words that might be local host names eg "fab" 

We may as well go for a complex address regex like https://stackoverflow.com/questions/53497/regular-expression-that-matches-valid-ipv6-addresses, which can be precompiled and will be quick enough to execute... ideally with a single expression rather than two.